sympy/core/add.py

```diff
@@ -1222,7 +1215,53 @@ def __neg__(self):
             return super().__neg__()
         return Add(*[-i for i in self.args])
 
-add = AssocOpDispatcher('add')
+
+UndefinedKind.add = Add
+NumberKind.add = Add
+
+
+def _add(args, start=0, evaluate=False, **kwargs):
+    # Function for sympy internal use. Intended to replace postprocessors.
+    # Introduced to avoid potential backwards incompatibility in the future.
+    # Note that this may be deleted if subclasses are removed.
+    return add(args, start=start, evaluate=evaluate, **kwargs)
```

Comment:
When I suggested the idea of `_add` I meant instead of having a public name `add`. If there is a public name `add` then I don't think that we also need a private `_add` function or perhaps it could be called `add_unevaluated` or something.

The reason I suggested this was because @Upabjojr seemed to be questioning whether adding a public function called `add` would be more useful for a different idea in future.

@Upabjojr what do you think of this PR?

It uses kind dispatching to provide functions `add` and `_add` that could be used to eliminate the need for post-processors in `Add` and provide a more systematic way for different types of object to extend the core arithmetic from outside of the core which I think is very useful.

The difference between `add` and `_add` here is that `add` defaults to `evaluate=True`making it a dropin replacement for `Add` but using the kind dispatching system. That means that calls to `Add` throughout the codebase could be changed to `add` to enable the new system.

Reply:
I think removing the postprocessor is better be done by removing subclasses.

In [previous comment](https://github.com/sympy/sympy/pull/20746#discussion_r554468552), you mentioned:
> We can't completely unify `MatAdd` and `Add` because they have different behaviour with respect to evaluation.

But this can be resolved by having different `NumberKind._eval_add` and `MatrixKind._eval_add`, so I expect that removing the subclasses can be easier than we think.

Reply:
OK, sounds good.

> I can't be certain for anything for now, so I'm planning to make another PR that actually unifies `Add` and `MatAdd`.

Great!